src/main/java/org/icij/datashare/Neo4jUtils.java

```diff
@@ -0,0 +1,722 @@
+            @JsonProperty("property") VariableProperty property,
+            @JsonProperty("direction") Objects.SortDirection direction
+        ) {
+            this.property = java.util.Objects.requireNonNull(property, "missing property");
+            this.direction = java.util.Objects.requireNonNull(direction, "missing sort direction");
+        }
+
+        @Override
+        public SortItem into() {
+            org.neo4j.cypherdsl.core.Property sortByProp = this.property.into();
+            if (this.direction.equals(Objects.SortDirection.ASC)) {
+                return sortByProp.ascending();
+            }
+            return sortByProp.descending();
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            SortByProperty that = (SortByProperty) o;
+            return java.util.Objects.equals(this.property, that.property)
+                && this.direction == that.direction;
+        }
+
+        @Override
+        public int hashCode() {
+            return java.util.Objects.hash(property, direction);
+        }
+    }
+
+    protected static class PatternNode implements Into<Node> {
+        protected final String name;
+        protected final List<String> labels;
+        protected final Map<String, Object> properties;
+
+        @JsonCreator
+        protected PatternNode(
+            @JsonProperty("name") String name,
+            @JsonProperty("labels") List<String> labels,
+            @JsonProperty("properties") Map<String, Object> properties
+        ) {
+            this.name = name;
+            this.labels = labels;
+            this.properties = properties;
+        }
+
+        @Override
+        public Node into() {
+            Node asNode;
+            if (this.labels == null || this.labels.isEmpty()) {
+                asNode = Cypher.anyNode();
+            } else {
+                asNode = Cypher.node(
+                    this.labels.get(0), this.labels.subList(1, this.labels.size()));
+            }
+            if (this.name != null) {
+                asNode = asNode.named(this.name);
+            }
+            if (this.properties != null && !this.properties.isEmpty()) {
+                asNode = asNode.withProperties(this.properties);
+            }
+            return asNode;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            PatternNode that = (PatternNode) o;
+            return java.util.Objects.equals(this.name, that.name)
+                && java.util.Objects.equals(this.labels, that.labels)
+                && java.util.Objects.equals(this.properties, that.properties);
```

Comment:
I have the feeling that comparison in the Java standard lib is correctly implemented for implementations of `Map<K, V>` as long as `K` and `V` properly implement `equals`.

So this relies on the fact that people will either use native objects for values `V` or correctly implement equals for their own objects. This is not ideal I agree.

A potential solution here would be to restrict have `properties : Map<K, V>` where `V` is restricted to native types instead of using `Objects` but I have no idea how to do so.

Reply:
Actually I think that as long as this class is used for deserialization we should be fine since Jackson will deserialize the `properties` as a Map of low level Java standard lib objects which should correctly implement `equals`

Reply:
Nice, thanks for looking up